### ISSUE TITLE
Question: Support for multiple endpoints in the same host service

### ISSUE THREAD

**Issue by AlexEngblom** [Particular/NServiceBus.Extensions.Hosting#27]
Is there support for multiple endpoints inside the same service? I suppose this might be an issue with shared DI and generic IMessageSession registrations inside the UseNServiceBus calls. Should we still use old approach of injecting dependencies separately through .RegisterComponents() to have better control over how endpoints are available for ie. Controllers.

**Comment by mauroservienti** [comment]
@AlexEngblom thanks for the question.

>Is there support for multiple endpoints inside the same service?

No, this package doesn't support multiple endpoints. Despite the fact that it'll allow you to create more than one by calling `.UseNServiceBus()` multiple times, you'll end up, as you correctly pointed out, with no control over the way things are registered into the container and which `IMessageSession` is then served to dependencies.

> Should we still use old approach of injecting dependencies separately through .RegisterComponents() to have better control over how endpoints are available for ie. Controllers.

yes, it's the optimal approach.

**Comment by steve-haar** [comment]
I ran across the same issue and modified the `UseNServiceBus` method to use its own service collection. I've made a fork here: https://github.com/vintage-software/NServiceBus.Extensions.Hosting.

The usage is like this:
```
return new HostBuilder()
    .ConfigureAppConfiguration(builder => GetConfiguration(builder))
    .UseNServiceBus(
        endpointConfigurationBuilder: hostBuilderContext => GetEndpointConfiguration1(),
        configureServicesForHandlers: ConfigureServices)
    .UseNServiceBus(
        endpointConfigurationBuilder: hostBuilderContext => GetEndpointConfiguration2(),
        configureServicesForHandlers: ConfigureServices);
```

**Comment by AlexEngblom** [comment]
@steve-haar maybe I'm missing something, but it seems like this solution doesn't address the main problem though. How can you be certain which endpoints IMessageSession is used when sending messages using your main service container context?

**Comment by steve-haar** [comment]
Are you talking about when you need to send a message outside of an NServiceBus handler? If that's the case, then you're right, with what I have the DI wouldn't be able to resolve IMessageSession. I forgot about that use case, because I didn't need to do that in my app.

I made a small modification and now if you resolve IMessageSession from your main container it will use whichever endpoint you set up last: https://github.com/vintage-software/NServiceBus.Extensions.Hosting/commit/393a410990868a1a2d08e111936b65dea2adeb6b

**Comment by timbussmann** [comment]
I wouldn't recommend your approach @steve-haar as the two endpoints can potentially share different internal services because there is only one service collection underneath.

at the moment you'd have to just host handlers for the different endpoints in a single endpoint. You can ship your handlers for different logical endpoints in separate assemblies and reference them all in the asp.net app, this will load all the handlers in the same endpoint. This shouldn't cause issues as long as you don't necessarily need multiple incoming queues for the same web app.

**Comment by steve-haar** [comment]
But with the changes I made here https://github.com/vintage-software/NServiceBus.Extensions.Hosting, I modified it so it actually does use a separate service collection for each endpoint underneath.

**Comment by timbussmann** [comment]
I see, the more important part is https://github.com/vintage-software/NServiceBus.Extensions.Hosting/commit/6f06d7f84b7c7923e4ae0c009903945b869a021d#diff-04a14c754811596fe9c906cec639846aR31 which ensures that every endpoint uses it's own service provider. As long as there is nothing you need to share between other hosted services and your endpoints, that should probably work at first glance.

**Comment by buvinghausen** [comment]
Put me down for wanting Particular to provide guidance and support for multiple endpoints. I can hear @udidahan telling me not to mistake deployment of services in the same process is not the same as coupling them.

**Comment by buvinghausen** [comment]
@timbussmann in the interim can I propose that we mark both NServiceBusHostedService & ServiceCollectionAdapter as public so that I can leverage them in my own IoC configuration?

**Comment by timbussmann** [comment]
> I can hear @udidahan telling me not to mistake deployment of services in the same process is not the same as coupling them.

that is right, there is also no word that it requires different service's handlers to be deployed to different endpoints on the same process though.

@buvinghausen you can easily copy over those two types into your project and make use of them as it fits your needs the best way. Marking internal types public means we'll have to properly deprecate them again once they should no longer be public and this process typically requires 2 major versions, I think you'd be much faster and flexible to take the parts which you require for your needs and adjust them accordingly.

**Comment by buvinghausen** [comment]
For anyone who lands here wanting to run multiple endpoints in the same host services here is the guidance I have come to after getting everything to work.

1. Do not use this library at all just go to [NServiceBus.Extensions.DependencyInjection](https://github.com/Particular/NServiceBus.Extensions.DependencyInjection)

2 . Copy the root IServiceCollection service descriptors to an array of ServiceDescriptors

```csharp
var serviceDescriptors = new ServiceDescriptor[services.Count];
services.CopyTo(serviceDescriptors, 0);
```

3. Configure the UseContainer extension method with an IServiceProviderFactory<IServiceCollection> instance and load the descriptors from step 2

```csharp
var containerSettings = endpointConfiguration.UseContainer(new DefaultServiceProviderFactory());
containerSettings.ConfigureContainer(sc =>
{
    foreach (var descriptor in serviceDescriptors)
        sc.Add(descriptor);
    // Additional service collection items for this endpoint instance go here
});
```

4. Create a custom hosted service class that takes in the dictionary of message sessions & the key you use to identify that endpoint so IMessageSession can get added when the hosted service starts up

```csharp
internal class CustomBusHostedService : IHostedService
{
    private IEndpointInstance _endpoint;
    private readonly IStartableEndpoint _startableEndpoint;
    private readonly IDictionary<string, IMessageSession> _messageSessions;
    private readonly string _key;

    public CustomBusHostedService(EndpointConfiguration endpointConfiguration, IDictionary<string, IMessageSession> messageSessions, string key)
    {
        _startableEndpoint = Endpoint.Create(endpointConfiguration).GetAwaiter().GetResult();
        _messageSessions = messageSessions;
        _key = key;
    }

    public async Task StartAsync(CancellationToken cancellationToken)
    {
        _endpoint = await _startableEndpoint.Start().ConfigureAwait(false);
        _messageSessions.Add(_key, _endpoint);
    }

    public Task StopAsync(CancellationToken cancellationToken) =>
        _endpoint.Stop();
    }
}
```

5. Add the hosted service to the root service collection

```csharp
services.AddSingleton<IHostedService>(new CustomBusHostedService(
    endpointConfiguration, endpointInstances, endpointName)); 
```

6. Add your message session dictionary to the main service collection to use in your controllers (or an additional provider if you want to make choosing the item out of the dictionary automatic (based on service namespace, etc.)

```csharp
services.AddSingleton<IDictionary<string, IMessageSession>>(endpointInstances);
```

Even though the services will spin up after the fact the memory reference to your original endpoint dictionary sticks around because it is a singleton so it hydrates correctly as they start.

**Comment by buvinghausen** [comment]
I fully realize this library is older than it's DI counterpart but if I were a maintainer at Particular I'd mark this guy an archive get the appropriate (tiny) files moved over and call it a day so that I wasn't maintaining two copies of ServiceCollectionAdapter and ServiceProviderAdapter and two NuGet packages that do largely the same things only the use case for this one is limited to one bus instance per host.

Cheers

**Comment by timbussmann** [comment]
> I fully realize this library is older than it's DI counterpart but if I were a maintainer at Particular I'd mark this guy an archive get the appropriate (tiny) files moved over and call it a day so that I wasn't maintaining two copies of ServiceCollectionAdapter and ServiceProviderAdapter and two NuGet packages that do largely the same things only the use case for this one is limited to one bus instance per host.

this package is fully supported and maintained. The split between the `NServiceBus.Extensions.DependencyInjection` and this package was intentional as they have different purposes:
* This package brings a convenient way to host NServiceBus with the generic host with a little amount of effort.
* The DI package brings support for the MS DI abstraction, similar to other DI container packages for NServiceBus.

While they share a tiny amount of code, they are basically intended for different use cases, as your alternative usage of the host shows nicely. The major difference in the usage scenario is the owner of the DI container. In the generic host case, the DI container lifecycle is managed by the host and shared with NServiceBus, while when using `NServiceBus.Extensions.DependencyInjection`, typically NServiceBus continues to own the container lifecycle (it is created and managed inside NServiceBus).

Thanks for sharing your approach to this so others can use this approach for now if they have the need to host multiple endpoints in the same generic host. If possible, I'd probably recommend using multiple hosts instead of multiple endpoints in a single host but it's difficult to make a general statement about this.

**Comment by buvinghausen** [comment]
I edited the hosted service in the example to allow generic key type definition as well as deferred the endpoint create call to the StartAsync method so that you get proper task handling rather than using GetAwaiter().GetResult() which always feels dirty.

> If possible, I'd probably recommend using multiple hosts instead of multiple endpoints in a single host but it's difficult to make a general statement about this.

We will have to agree to disagree there I have two hosts one for web and one for worker but all services get discovered as plugins and configured accordingly.

I could presumably run a meta bus instance for all services in both but that comes with some pretty serious issues on the worker side as the worker role handles persistence.  I want each service to have it's own persistence setup so therefore I must have multiple endpoints running.  Once I'm running multiple endpoints on the worker side it would be silly to have the web run in a different configuration even though it's far more feasible over there because I can route each command assembly to the different running worker instances.

Yes I could presumably split every single web & worker into separate hosts but then if I want to add a service to the ecosystem now I'm having to go create kubernetes pods for the new service at every level go work on service discovery setup when it's just much easier for my two hosts to discover the new plugin and start running.  I find the developer experience is much greater too.  Telling them to run three host processes (webpack dev server + worker + web) is much easier than trying to figure out which service mix they are working with and what's needed to support what they are working on.

**Comment by danielmarbach** [comment]
TLDR:

- We are not going to enable multi hosting as part of this extension
- The new version of this extension will now prevent accidentally trying to use this library for multi-hosting to avoid overwriting service collection registrations

Here is an explanation of why

## The host, a unit of encapsulation prevents multi-hosting

By design, the Generic host assumes it is the [unit of encapsulation](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?#host-definition) for the application or service hosted. All application resources applied to the same host instance share a single service collection and, therefore a single service provider. The conclusion of these design assumptions is that a generic host can only host a single service of a specific type by default. So, for example, you can host ASP.NET Core WebApi together with NServiceBus, but it wouldn't be possible to host multiple ASP.NET Core WebApis or NServiceBus instances as part of the generic host without doing advanced trickery like [overriding controller detection for WebApi](https://www.strathweb.com/2017/04/running-multiple-independent-asp-net-core-pipelines-side-by-side-in-the-same-application/) or heavily [customizing assembly scanning](https://docs.particular.net/nservicebus/hosting/assembly-scanning) and more for NServiceBus. What both NServiceBus and WebApi have in common that by default, the assume to own all assemblies found in the application domain for convenience and ease of use.

Still, once you have experienced the power of `UseNServiceBus` you might be tempted to use it to host multiple NServiceBus instances on the same host like the following.

```
var host = Host.CreateDefaultBuilder()
   .UseNServiceBus(hostBuilderContext =>
      {
         var endpointConfiguration = new EndpointConfiguration("MyEndpoint");
         // ...
         return endpointConfiguration;
   })
   .UseNServiceBus(hostBuilderContext =>
      {
         var endpointConfiguration = new EndpointConfiguration("MyOtherEndpoint");
         // ...
         return endpointConfiguration;
   })   
```

In previous versions of the extension, this could not be prevented from being executed and would have lead to obstructive runtime behavior like the last endpoint configuration being used overriding the previous configuration. With this release it a proper safeguard has been put into place that warns about this improper usage.

> `UseNServiceBus` can only be used once on the same host instance because subsequent calls would override each other. [...]

With the broad adoption of continuous deployment pipelines (i.ex. Github Actions, Octopus Deploy...) and service orchestration mechanisms like Kubernetes, Service Fabric, Docker Compose and many more it is encouraged to have one generic host per NServiceBus instance. Multiple endpoints sharing the same generic host drastically increases the configuration complexity and couples those endpoints together, for example if one endpoint has high CPU/memory consumption, the other hosted in the same process might starve. These are the reasons it is actively prevented from doing `UseNServiceBus` on the same host, and we are not planning to implement the support for multiple endpoints into the extension.

Should you still have good reasons to host multiple endpoints in the same generic host, it is recommended to use multiple generic hosts to achieve proper isolation. It is still required to configure assembly scanning carefully. [Here is a sample](https://docs.particular.net/samples/hosting/generic-multi-hosting/) showing it how to host multip endpoints using multiple generic host instances. We encourage you to always use this approach for multi hosting because it provides the necessary isolation from a DI perspective. 

Should you still have needs to share common dependencies (and I would first think multiple times about different solutions before going down this path) I've put together a [draft PR that shows how the collection can be shared](https://github.com/Particular/docs.particular.net/pull/4813).
